src/vm/hints/sha256_utils.rs

```diff
@@ -0,0 +1,242 @@
+            bigint_to_u32, get_integer_from_var_name, get_ptr_from_var_name,
+            get_relocatable_from_var_name, insert_value_from_var_name,
+        },
+        vm_core::VirtualMachine,
+    },
+};
+
+use generic_array::GenericArray;
+use num_bigint::BigInt;
+use num_traits::{One, Zero};
+use sha2::compress256;
+use std::collections::HashMap;
+
+const SHA256_INPUT_CHUNK_SIZE_FELTS: usize = 16;
+const SHA256_STATE_SIZE_FELTS: usize = 8;
+const BLOCK_SIZE: usize = 7;
+
+pub fn sha256_input(
+    vm: &mut VirtualMachine,
+    ids: &HashMap<String, BigInt>,
+    hint_ap_tracking: Option<&ApTracking>,
+) -> Result<(), VirtualMachineError> {
+    let n_bytes = get_integer_from_var_name(
+        "n_bytes",
+        ids,
+        &vm.memory,
+        &vm.references,
+        &vm.run_context,
+        hint_ap_tracking,
+    )?;
+
+    insert_value_from_var_name(
+        "full_word",
+        if n_bytes >= &bigint!(4) {
+            BigInt::one()
+        } else {
+            BigInt::zero()
+        },
+        ids,
+        &mut vm.memory,
+        &vm.references,
+        &vm.run_context,
+        hint_ap_tracking,
+    )
+}
+
+pub fn sha256_main(
+    vm: &mut VirtualMachine,
+    ids: &HashMap<String, BigInt>,
+    hint_ap_tracking: Option<&ApTracking>,
+) -> Result<(), VirtualMachineError> {
+    let sha256_start = get_relocatable_from_var_name(
+        "sha256_start",
+        ids,
+        &vm.memory,
+        &vm.references,
+        &vm.run_context,
+        hint_ap_tracking,
+    )?;
+
+    let input_ptr = vm.memory.get_relocatable(&sha256_start)?;
+
+    let mut message: Vec<u8> = Vec::with_capacity(4 * SHA256_INPUT_CHUNK_SIZE_FELTS);
+
+    for i in 0..SHA256_INPUT_CHUNK_SIZE_FELTS {
+        let input_element = vm.memory.get_integer(&(input_ptr + i))?;
+        let bytes = bigint_to_u32(input_element)?.to_be_bytes();
+        message.extend(bytes);
+    }
+
+    let mut iv: [u32; SHA256_STATE_SIZE_FELTS] = [
+        0x6A09E667, 0xBB67AE85, 0x3C6EF372, 0xA54FF53A, 0x510E527F, 0x9B05688C, 0x1F83D9AB,
+        0x5BE0CD19,
+    ];
+
+    let new_message = GenericArray::clone_from_slice(&message);
+    compress256(&mut iv, &[new_message]);
+
+    let mut output: Vec<BigInt> = Vec::with_capacity(SHA256_STATE_SIZE_FELTS);
+
+    for new_state in iv {
+        output.push(bigint!(new_state));
+    }
+
+    let output_base = get_ptr_from_var_name(
+        "output",
+        ids,
+        &vm.memory,
+        &vm.references,
+        &vm.run_context,
+        hint_ap_tracking,
+    )?;
+
+    vm.segments
+        .write_arg(&mut vm.memory, &output_base, &output, Some(&vm.prime))
+        .map_err(VirtualMachineError::MemoryError)?;
+    Ok(())
+}
+
+pub fn sha256_finalize(
+    vm: &mut VirtualMachine,
+    ids: &HashMap<String, BigInt>,
+    hint_ap_tracking: Option<&ApTracking>,
+) -> Result<(), VirtualMachineError> {
+    let message: Vec<u8> = vec![0; 64];
+
+    let mut iv: [u32; SHA256_STATE_SIZE_FELTS] = [
```

Comment:
This IV is repeated across functions, could it be defined as a constant?

Reply:
Resolved: https://github.com/lambdaclass/cleopatra_cairo/pull/357/commits/8e5f9cb27eb994fe78e45ab63a8126ee3008d099